cmd/customd/client/client.go

```diff
@@ -0,0 +1,448 @@
+
+	uuid := hex.EncodeToString(append(tx.Hash(), cmn.RandBytes(2)...))
+	evts, err := cc.conn.Subscribe(ctx, uuid, query.String())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to subscribe")
+	}
+
+	// make sure to unregister after the test is over
+	defer cc.conn.UnsubscribeAll(ctx, uuid)
+
+	select {
+	case evt := <-evts:
+		return evt.Data.(tmtypes.TMEventData), nil
+	case <-ctx.Done():
+		return nil, errors.New("timed out waiting for event")
+	}
+}
+
+// BroadcastTxAsync can be run in a goroutine and will output
+// the result or error to the given channel.
+// Useful if you want to send many tx in parallel
+func (cc *CustomClient) BroadcastTxAsync(tx weave.Tx, out chan<- BroadcastTxResponse) {
+	data, err := tx.Marshal()
+	if err != nil {
+		out <- BroadcastTxResponse{Error: err}
+		return
+	}
+
+	// TODO: make this async, maybe adjust return value
+	res, err := cc.conn.BroadcastTxCommit(data)
+	msg := BroadcastTxResponse{
+		Error:    err,
+		Response: res,
+	}
+	out <- msg
+}
+
+// SubscribeHeaders queries for headers and starts a goroutine
+// to typecase the events into Headers. Returns a cancel
+// function. If you don't want the automatic goroutine, use
+// Subscribe(QueryNewBlockHeader, out)
+func (cc *CustomClient) SubscribeHeaders(out chan<- *Header) (func(), error) {
+	query := QueryNewBlockHeader
+	pipe, cancel, err := cc.Subscribe(query)
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		for msg := range pipe {
+			evt, ok := msg.Data.(tmtypes.EventDataNewBlockHeader)
+			if !ok {
+				// TODO: something else?
+				panic("Unexpected event type")
+			}
+			out <- &evt.Header
+		}
+		close(out)
+	}()
+	return cancel, nil
+}
+
+// Subscribe will take an arbitrary query and push all events to
+// the given channel. If there is no error,
+// returns a cancel function that can be called to cancel
+// the subscription
+func (cc *CustomClient) Subscribe(query tmpubsub.Query) (<-chan ctypes.ResultEvent, func(), error) {
+	ctx := context.Background()
+	out, err := cc.conn.Subscribe(ctx, cc.subscriber, query.String())
+	if err != nil {
+		return out, nil, err
+	}
+	cancel := func() {
+		cc.conn.Unsubscribe(ctx, cc.subscriber, query.String())
+	}
+	return out, cancel, nil
+}
+
+// UnsubscribeAll cancels all subscriptions
+func (cc *CustomClient) UnsubscribeAll() error {
+	ctx := context.Background()
+	return cc.conn.UnsubscribeAll(ctx, cc.subscriber)
+}
+
+// GetWallet will return a wallet given an address
+// If non wallet is present, it will return (nil, nil)
+// Error codes are used when the query failed on the server
+func (cc *CustomClient) GetWallet(addr weave.Address) (*WalletResponse, error) {
+	// make sure we send a valid address to the server
+	err := addr.Validate()
+	if err != nil {
+		return nil, errors.WithMessage(err, "Invalid Address")
+	}
+
+	resp, err := cc.AbciQuery("/wallets", addr)
+	if err != nil {
+		return nil, err
+	}
+	if len(resp.Models) == 0 { // empty list or nil
+		return nil, nil
```

Comment:
I think if there is no result, `errors.ErrNotFound` should be returned instead of `nil`.

This could be applied to all query wrappers.